artmanflow/steps/common.py

```diff
@@ -0,0 +1,367 @@
+            pass
+
+    def run_guest_script(self, guest_config, extra_mount_volumes=None):
+        host_guest_config_file_path = self._host.host_guest_config_file_path()
+        ConfigUtils.dump_config(guest_config, host_guest_config_file_path)
+
+        host_guest_output_dir = self._host.host_guest_output_dir_path()
+        os.makedirs(host_guest_output_dir)
+
+        mount_volumes = [
+            [self._host.step_dir_path(), self._guest.guest_step_dir_path(),
+             'ro'],
+            [host_guest_output_dir, self._guest.guest_output_dir_path()],
+        ]
+
+        if extra_mount_volumes:
+            mount_volumes.extend(extra_mount_volumes)
+
+        script_path = self._guest.guest_script_path(
+            self._host.guest_script_name())
+        self._run_command(['python3', script_path],
+                          host_guest_config_file_path,
+                          self._host.stdout_file_path(),
+                          self._guest.guest_root_path(),
+                          mount_volumes)
+
+    def _run_command(self, command, input_file_name, output_file_name,
+        guest_root_path=None, mount_dirs=None):
+
+        cmd = self._construct_docker_run_command(command, guest_root_path,
+                                                 mount_dirs)
+        with open(input_file_name, 'r') as input_file, \
+            open(output_file_name, 'a+') as output_file:
+            output_file.write(
+                "\033[1;30msubprocess.Popen(%s, stdout='%s', stderr='%s', stdin='%s'\n\033[0m\n" % (
+                    cmd, output_file_name, output_file_name, input_file_name))
+            output_file.flush()
+            subprocess.Popen(cmd, stdout=output_file, stderr=output_file,
+                             stdin=input_file)
+
+    def _construct_docker_run_command(self, command, guest_root_path=None,
+        mount_dirs=None):
+        cmd = ['docker', 'run', '--rm']  # run and remove container after exit
+        if guest_root_path:
+            cmd += ['-w', guest_root_path]  # set working dir
+        if mount_dirs:
+            for mount_dirs_pair in mount_dirs:
+                cmd += ['-v', ':'.join(mount_dirs_pair)]  # mount volumes
+        cmd += [
+            '-e', 'HOST_USER_ID=%s' % os.getuid(),  # to chown inside  docker
+            '-e', 'HOST_GROUP_ID=%s' % os.getgid(),
+            '-i', self._config['docker_image']  # specify image
+        ]
+        cmd += command
+
+        return cmd
+
+    def local_repo_mount(self, config):
+        repo = config['git_repo']
+        if repo.startswith('file://'):
+            local_mount_path = repo[len('file://'):]
+        elif repo.startswith('/'):
+            local_mount_path = repo
+        else:
+            return []
+
+        local_repo_name = local_mount_path[local_mount_path.rfind('/') + 1:]
+        guest_mount_path = self._guest.guest_root_subpath(local_repo_name)
+
+        return [local_mount_path, guest_mount_path]
+
+
+class BaseGuest:
```

Comment:
Naming: "guest" seems to signify the context of the work being performed; potentially a purpose-based name would be better? For example, the term "guest" doesn't quite make sense when running in standalone mode.